java/test/com/facebook/watchman/WatchmanConnectionTest.java

```diff
@@ -295,5 +297,83 @@ public void onFailure(Throwable t) {
+        };
+    Callable<Map<String, Object>> incomingMessageGetter = new Callable<Map<String, Object>>() {
+      @Override
+      public Map<String, Object> call() throws Exception {
+        commandSentSemaphore.acquire();
+        BserDeserializer.BserEofException e = new BserDeserializer.BserEofException("expected");
+        expectedExceptionRef.set(e);
+        throw e;
+      }
+    };
+    WatchmanConnection connection = new WatchmanConnection(
+        incomingMessageGetter,
+        mOutgoingMessageStream,
+        Optional.<WatchmanConnection.WatchmanCommandListener>of(commandListener));
+    connection.start();
+    ListenableFuture<Map<String, Object>> listenableFuture = connection.run(Arrays.asList("clock", "/a/b/c"));
+    try {
+      Map<String, Object> receivedResponse = listenableFuture.get(5, TimeUnit.SECONDS);
+      Assert.fail("Should not get a response, got " + receivedResponse.toString());
+    } catch (ExecutionException executionException) {
+      Assert.assertSame(expectedExceptionRef.get(), executionException.getCause());
+    }
+  }
+
+  @Test
+  public void propagatesInterruptionException() throws Exception {
+    final Semaphore enteredCallable = new Semaphore(0);
+    final Semaphore exitedCallable = new Semaphore(0);
+    final AtomicReference<Thread> threadRef = new AtomicReference<>();
+    final AtomicReference<InterruptedException> expectedException = new AtomicReference<>();
+    Callable<Map<String, Object>> incomingMessageGetter = new Callable<Map<String, Object>>() {
+      @Override
+      public Map<String, Object> call() throws Exception {
+        try {
+          threadRef.set(Thread.currentThread());
+          enteredCallable.release();
+          Thread.sleep(1);
+          int i = 0;
+          while (true) {
+            i++;
+          }
+        } catch (InterruptedException e) {
+          expectedException.set(e);
+          throw e;
+        } finally {
+          exitedCallable.release();
+        }
+      }
+    };
+    WatchmanConnection connection = new WatchmanConnection(incomingMessageGetter, mOutgoingMessageStream);
+    connection.start();
+    ListenableFuture<Map<String, Object>> listenableFuture = connection.run(Arrays.asList("clock", "/a/b/c"));
+    enteredCallable.acquire();
+    threadRef.get().interrupt();
+    Assert.assertTrue("Callable should have exited", exitedCallable.tryAcquire(5, TimeUnit.SECONDS));
+    try {
+      listenableFuture.get(5, TimeUnit.SECONDS);
+      Assert.fail("Should not be able to get a result from an interrupted operation");
+    } catch (InterruptedException e) {
+      // This is reasonable behavior...
+      Assert.assertSame("If InterruptedException, should be the one thrown by thread",
+          expectedException.get(), e);
+    } catch (ExecutionException e) {
+      // ...and this is also reasonable behavior.
+      Assert.assertSame("If ExecutionException, cause should be exception that interrupted thread",
```

Comment:
This double-catch is weird, but reading the documentation for the interface method `Future#get(int,TimeUnit)`, it seems to be up to the implementation's discretion whether it throws a bare `InterruptedException` or wraps it in an `ExecutionException`.  

Since both are arguably reasonable interpretations (and since the most likely user case is for the caller to generate a stack trace or print the message, in which case these are equally useful), as long as it's one or the other, that seems reasonable to me.